### PR TITLE
Fix strategy checkpoint restore order for model architecture hooks

### PR DESCRIPTION
**Pull request by ys-teh**
# ALCHEMI Toolkit Pull Request

## Description

Fix full strategy checkpoint restore to address #132 so registration-time model architecture hooks are applied before model weights and optimizer state are restored. This preserves restart semantics for fine-tuning module patches and adapter-style hooks whose modules must exist before loading checkpoint state.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

N/A

## Changes Made

- Rebuild full strategy checkpoints by constructing unweighted models, registering strategy hooks, then restoring model, optimizer, scheduler, runtime, and hook state.
- Add a regression test covering `FineTuningStrategy.module_patches` that replace an existing module and add a new module before checkpoint state restore.

## Testing

Targeted tests run:

- `uv run pytest test/training/test_checkpoint.py` (passed, 62 tests)
- `uv run pytest test/training/test_finetune.py` (passed, 20 tests)
- `uv run ruff check nvalchemi/training/_checkpoint.py test/training/test_checkpoint.py` (passed)

- [ ] Unit tests pass locally (`make pytest`)
- [ ] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

When `model_names` is provided, loading remains component-level. For checkpoints whose model architecture comes from hooks, users should load the full strategy and extract the model.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: test/training/test_checkpoint.py [hunk 1]
```diff
@@ -1319,6 +1319,71 @@ def test_strategy_checkpoint_loads_builtin_result_and_restart_state(
...
+        optimizer_params = {
+            id(parameter)
+            for group in restored._optimizers[0].param_groups
+            for parameter in group["params"]
+        }
+        assert id(restored_model.projection.weight) in optimizer_params
+        assert id(restored_model.aux_projection.weight) in optimizer_params
+        assert any(
+            parameter is restored_model.projection.weight
+            for parameter in restored._optimizers[0].state
```
[the remark this answers is not in the document]
Comment (ys-teh) [reply test/training/test_checkpoint.py]: Ended up simplifying the test to remove .projection altogether since it is not the main target.

File: nvalchemi/training/_checkpoint.py [hunk 2]
```diff
@@ -1402,6 +1402,32 @@ def _with_strategy_device_override(
...
 
+def _build_model_from_checkpoint_spec(
+    root: Path,
+    name: str,
+    *,
+    load_location: torch.device | None,
+) -> tuple[nn.Module, BaseSpec]:
+    """Build a model without its saved weights from its checkpoint spec."""
+    spec = _load_spec(root / "models" / name / "spec.json")
+    build_kwargs = (
```
Comment (laserkelvin) [inline nvalchemi/training/_checkpoint.py]: Should we expand this to include more possible build kwargs? We could just splat `**kwargs` into here

File: nvalchemi/training/_checkpoint.py [hunk 3]
```diff
@@ -1402,6 +1402,32 @@ def _with_strategy_device_override(
...
+        {"device": load_location}
+        if load_location is not None and spec.accepts_kwarg("device")
+        else {}
+    )
+    model = spec.build(**build_kwargs)
+    if not isinstance(model, nn.Module):
```
Comment (laserkelvin) [inline nvalchemi/training/_checkpoint.py:1428]: I understand why we have this, but I'm not 100% sure we should gate it like this in case for whatever reason, someone brings a model along that is not best practice and it's actually some container that contains a `nn.Module`. What do you think about this?

File: nvalchemi/training/_checkpoint.py [hunk 4]
```diff
@@ -1692,6 +1718,44 @@ def load_checkpoint(
         _run_validators(loaded, validators)
         return loaded
 
+    if strategy_metadata is not None and model_names is None:
+        from nvalchemi.training.strategy import TrainingStrategy
+
+        unweighted_models = {
```
Comment (laserkelvin) [inline nvalchemi/training/_checkpoint.py:1739]: Some short, inline comments sprinkled here would be good to explain what is being done and why

File: test/training/test_checkpoint.py [hunk 5]
```diff
@@ -1319,6 +1319,58 @@ def test_strategy_checkpoint_loads_builtin_result_and_restart_state(
         assert restored.step_count == 4
         assert restored.global_step_count == 4
 
+    def test_finetuning_checkpoint_loads_after_registration_patches(
```
Comment (laserkelvin) [inline test/training/test_checkpoint.py:1323]: Parameterize this test to include GPU as well.

If you want to keep the exception being raised for when the "model" is not an `nn.Module`, probably add a test for that branch as well

**Review by laserkelvin — CHANGES_REQUESTED** [review]
Idea and implementation looks good to me, just added some discussion points and some minor asks

File: test/training/test_checkpoint.py (hunk 5 above)
Comment (ys-teh) [reply test/training/test_checkpoint.py:1323]: I added tests for GPU and fixed issues related to device placement.

File: nvalchemi/training/_checkpoint.py (hunk 4 above)
Comment (ys-teh) [reply nvalchemi/training/_checkpoint.py:1739]: Updated.

**Comment by laserkelvin** [comment]
/ok to test [f52f419](https://github.com/NVIDIA/nvalchemi-toolkit/pull/133/commits/f52f419a61da435494264d41e99cc751b89c038f)

File: nvalchemi/training/_checkpoint.py [hunk 6]
```diff
@@ -1697,6 +1706,9 @@ def load_checkpoint(
     )
     load_location = _strategy_target_device(strategy_metadata, map_location)
 
+    # Path 1: hydrate a user-supplied live strategy.
```
Comment (laserkelvin) [inline nvalchemi/training/_checkpoint.py]: Don't use the word hydrate XD

Comment (ys-teh) [reply nvalchemi/training/_checkpoint.py]: Changed to restore

File: nvalchemi/training/_checkpoint.py (hunk 3 above)
Comment (ys-teh) [reply nvalchemi/training/_checkpoint.py:1428]: Added a comment on why nn.Module is required. We likely need to detect it when saving, not just checking when loading.

**Review by laserkelvin — APPROVED** [review]
Good to go on my end!

**Comment by laserkelvin** [comment]
/ok to test [96bd226](https://github.com/NVIDIA/nvalchemi-toolkit/pull/133/commits/96bd22672f643230282497183d473ba548ea2047)

**Comment by ys-teh** [comment]
/ok to test [45a9cdc](https://github.com/NVIDIA/nvalchemi-toolkit/pull/133/commits/45a9cdcd3999c85fe11839dc1e8ec0bb41df5e62)
